prov/gni: disallow FI_REMOTE_CQ_DATA w/FI_MORE

a-abraham opened the pull request:
FI_MORE cannot currently support FI_REMOTE_CQ_DATA.
This commit will disallow requests with said flag from
being added to the fi_more list.

fixes ofi-cray/libfabric-cray#1110

Signed-off-by: Amith Abraham <aabraham@cray.com>

sungeunchoi commented:
Requesting test program.

hppritcha commented:
@a-abraham will have Time to add test?

ztiffany commented on prov/gni/src/gnix_rma.c:
```diff
@@ -1352,8 +1352,10 @@ ssize_t _gnix_rma(struct gnix_fid_ep *ep, enum gnix_fab_req_type fr_type,
+	 * Current FI_MORE implementation does not create remote CQ events.
+	 * Remove FI_MORE flag when FI_FENCE or FI_REMOTE_CQ_DATA flags are
+	 * present. */
+	if ((flags & FI_FENCE) || (flags & FI_REMOTE_CQ_DATA)) {
```
I think you also need to check for:

req->vc->peer_caps & FI_RMA_EVENT

In this case, we send an event to the target to increment a remote counter.  We don't support that with FI_MORE.

Once that's done you should be able to remove the 'fi_more_set' workaround from our criterion tests.

a-abraham replied:
Just rebased and i'm not seeing any issues. 

However I tried the following change and some corresponding code to remove this case from being queued on the FI_MORE list
```
- if ((flags & FI_FENCE) || (flags & FI_REMOTE_CQ_DATA)) {
+       if ((flags & FI_FENCE) || (flags & FI_REMOTE_CQ_DATA) ||
+           (ep->vc->peer_caps & FI_RMA_EVENT)) {
                flags &= ~FI_MORE;
        }
```
But I'm observing failures on a lot of gnitests afterwards. Trying to figure out why that might be happening...

a-abraham replied:
This has now been added and works successfully.

ztiffany commented on prov/gni/src/gnix_rma.c:
```diff
@@ -1452,14 +1446,28 @@ ssize_t _gnix_rma(struct gnix_fid_ep *ep, enum gnix_fab_req_type fr_type,
+	 * the fi_more list. */
+	if ((flags & FI_MORE) ||
+	    (!(flags & FI_FENCE) && !(flags & FI_REMOTE_CQ_DATA) &&
+	     !(req->vc->peer_caps & FI_RMA_EVENT) &&
```
Aren't these checks redundant?

ztiffany commented on prov/gni/src/gnix_rma.c:
```diff
@@ -1452,14 +1446,28 @@ ssize_t _gnix_rma(struct gnix_fid_ep *ep, enum gnix_fab_req_type fr_type,
+	 * the fi_more list. */
+	if ((flags & FI_MORE) ||
+	    (!(flags & FI_FENCE) && !(flags & FI_REMOTE_CQ_DATA) &&
+	     !(req->vc->peer_caps & FI_RMA_EVENT) &&
```
aren't these two lines just:

!(flags & FI_MORE)

so:

```
if (has_more || (!has_more && has_more_reqs))
    add to more list
```

a-abraham replied:
That is the general idea - however it's a little more nuanced right?
```
if (has_more ||  
     (!has_more  && has_more_reqs && !(FI_FENCE || FI_REMOTE_CQ_DATA || FI_RMA_EVENT)) 
      add to more list
```
I do think I will edit this to make it more readable. Let me know if I've missed the point of your comment though

ztiffany replied:
that code is different than what we have.  I see this (added parens):

if (has_more ||  
     (!has_more  && **(** has_more_reqs && !(FI_FENCE || FI_REMOTE_CQ_DATA || FI_RMA_EVENT) **)** ) 
      add to more list

a-abraham replied:
If you distribute the && through the parens that's the same as above right?
That being said, like I mentioned it is messy and I'm going to update that real quick.

a-abraham replied:
How does this logic look to you @ztiffany   checks out with what we discussed above?
Formatting it well is weird though
```
 if ((flags & FI_MORE) ||
     (!(flags & FI_MORE) &&
     (!slist_empty(&ep->more_write) || !slist_empty(&ep->more_read)) &&
     !(flags & FI_FENCE || flags & FI_REMOTE_CQ_DATA || req->vc->peer_caps & FI_RMA_EVENT)))
```

ztiffany commented on prov/gni/src/gnix_rma.c:
```diff
@@ -1497,9 +1505,10 @@ ssize_t _gnix_rma(struct gnix_fid_ep *ep, enum gnix_fab_req_type fr_type,
+		/* Requests with FI_FENCE or REMOTE EP requirements are not
+		 * added to the FI_MORE List. They must be queued separately. */
+		if ((flags & FI_FENCE) || (flags & FI_REMOTE_CQ_DATA) ||
+		    (req->vc->peer_caps & FI_RMA_EVENT)) {
```
Same as above.  Could this be:
```
if (!(flags & FI_MORE))
```

a-abraham replied:
Wouldn't this capture the first message without FI_MORE (which we don't want to queue separately) ?

ztiffany replied:
Queued separately, ok makes sense now.

ztiffany commented on prov/gni/src/gnix_rma.c:
```diff
@@ -1452,14 +1446,28 @@ ssize_t _gnix_rma(struct gnix_fid_ep *ep, enum gnix_fab_req_type fr_type,
+		flags &= ~FI_MORE;
+	}
+
+
```
Multiple blank lines added.

ztiffany commented:
I think I get it now.  It's pretty complicated since we try to treat "more-able" requests without FI_MORE and "non-more-able" requests without FI_MORE differently

a-abraham commented:
@ztiffany  I've added the code that disallows FI_MORE prior to a connection having been established, and have also updated the tests. One last review might be good, but I think we're done here.

ztiffany commented:
@a-abraham looks good.
